### ISSUE TITLE
5.2 RUNUAT with Vs 2019 on this plugin

### ISSUE THREAD

**Issue by aknittel1** [microdee/UE4-SpaceMouse#90]
Is it possible to just RUNUAT with the new 5.2 to generate your plugin, assuming vs 2019 is installed, not VS 2020? I ve domne this in the past with 5.0 and 5.1 releases.

**Comment by microdee** [comment]
yes it is

**Comment by Hunanbean** [comment]
Oh, perhaps that is why i could not update it. i was trying to use vs2022, and just erroring out

**Comment by aknittel1** [comment]
No good. I've generated this for 5.0, and 5.1. Fails with 5.2:

[18/21] Compile [x64] Module.SpaceMouse.cpp
D:\Data\Downloads\Browser\SpaceMouse-Win64-1.2.4.5.2.0-Source\SpaceMouse510Prebuild\HostProject\Plugins\SpaceMouse\Source\SpaceMouse\Public\SSmKeySelector.h(43): warning C4996: 'FEditorStyle::GetFontStyle': FEditorStyle::GetFontStyle() is deprecated, use FAppStyle::GetFontStyle() instead. Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile.
D:\Data\Downloads\Browser\SpaceMouse-Win64-1.2.4.5.2.0-Source\SpaceMouse510Prebuild\HostProject\Plugins\SpaceMouse\Source\SpaceMouse\Private\SmViewportOverlay.cpp(41): error C2065: 'FSceneViewFamilyContext': undeclared identifier
D:\Data\Downloads\Browser\SpaceMouse-Win64-1.2.4.5.2.0-Source\SpaceMouse510Prebuild\HostProject\Plugins\SpaceMouse\Source\SpaceMouse\Private\SmViewportOverlay.cpp(41): error C2146: syntax error: missing ';' before identifier 'ViewFam'
D:\Data\Downloads\Browser\SpaceMouse-Win64-1.2.4.5.2.0-Source\SpaceMouse510Prebuild\HostProject\Plugins\SpaceMouse\Source\SpaceMouse\Private\SmViewportOverlay.cpp(41): error C3861: 'ViewFam': identifier not found
D:\Data\Downloads\Browser\SpaceMouse-Win64-1.2.4.5.2.0-Source\SpaceMouse510Prebuild\HostProject\Plugins\SpaceMouse\Source\SpaceMouse\Private\SmViewportOverlay.cpp(47): error C2065: 'ViewFam': undeclared identifier
D:\Data\Downloads\Browser\SpaceMouse-Win64-1.2.4.5.2.0-Source\SpaceMouse510Prebuild\HostProject\Plugins\SpaceMouse\Source\SpaceMouse\Private\SmViewportOverlay.cpp(50): error C3536: 'SceneView': cannot be used before it is initialized
D:\Data\Downloads\Browser\SpaceMouse-Win64-1.2.4.5.2.0-Source\SpaceMouse510Prebuild\HostProject\Plugins\SpaceMouse\Source\SpaceMouse\Private\SSmKeySelector.cpp(192): warning C4996: 'FEditorStyle::GetBrush': FEditorStyle::GetBrush() is deprecated, use FAppStyle::GetBrush() instead. Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile.
D:\Data\Downloads\Browser\SpaceMouse-Win64-1.2.4.5.2.0-Source\SpaceMouse510Prebuild\HostProject\Plugins\SpaceMouse\Source\SpaceMouse\Private\SSmKeySelector.cpp(567): warning C4996: 'FEditorStyle::GetBrush': FEditorStyle::GetBrush() is deprecated, use FAppStyle::GetBrush() instead. Please update your code to the new API before upgrading to the next release, otherwise your project will no longer compile.
[19/21] Link [x64] UnrealEditor-SpaceMouse.dll cancelled
[20/21] Link [x64] UnrealEditor-SpaceMouse.lib cancelled
[21/21] WriteMetadata UnrealEditor.target cancelled
Took 78.569913s to run dotnet.exe, ExitCode=6
UnrealBuildTool failed. See log for more details. (C:\Users\aknit\AppData\Roaming\Unreal Engine\AutomationTool\Logs\C+Program+Files+Epic+Games+UE_5.2\UBT-UnrealEditor-Win64-Development.txt)
AutomationTool executed for 0h 1m 21s
AutomationTool exiting with ExitCode=6 (6)
BUILD FAILED

**Comment by microdee** [comment]
Thanks for figuring that out I thought they changed API's. I'll make an update once I'm home again.

**Comment by anttin2020** [comment]
Thank you @aknittel1 , I have no experience in c++ coding but I just added that line to that file and ran runuat again, now SpaceMouse works in UE 5.2 :)

**Comment by enamodeka** [comment]
Would you mind pointing me in the direction where I could understand how I can build this pluging from source using `RunUAT`?


> I worked this out. In your source , in file SmViewportOverlay.cpp, you need to add the line
> 
> #include "SceneView.h"

**Comment by anttin2020** [comment]
I succeeded by doing this :)
NOTE: I used this method for paths but there are other ways to do it as well, anyway this method worked for me.

This procedure needs Visual Studio 2019 but you just install it, no need to learn to use it :)
See also this: https://dev.epicgames.com/community/learning/tutorials/0RG4/compiling-a-plugin-for-a-new-engine-version.

1. Copy old, working plugin folder from UE 5.1 to UE 5.2 (paths below)
2. Edit (e.g. notepad++) the source file in the copy as instructed in this thread (add the include after first include in that file)
3. Open Command Prompt and go to folder (naturally adjust for your system) S:\UE\UE_5.2\Engine\Build\BatchFiles
    run command line:
    RunUAT.bat BuildPlugin -plugin="s:\ue\ue_5.2\engine\plugins\spacemouse\spacemouse.uplugin" -package="s:\newplugin" -TargetPlatforms=Win64
    where
    -"s:\ue\ue_5.2\engine\plugins\spacemouse\spacemouse.uplugin" = path to old plugin (the copied one so that edited source inside it will be in the new version as well)
    -"s:\newplugin" = where the new plugin folder tree will be created. This can be anywhere as the final plugin folder tree will be copied from here to the -plugin folder. But you CAN'T use the same path as -package. Clearest to me was to use totally separate folder but you do as you like. Anyway this way the modified sources are included in the new plugin folder tree which is handy for next update :)
4. Clear the contents of the plugin folder in UE 5.2 and copy the content of s:\newplugin there

Now start some project, activate SpaceMouse-plugin, all done!

**Comment by enamodeka** [comment]
@anttin2020 Wonderful answer. Million thanks! I'm gonna give it a go.

**Comment by anttin2020** [comment]
> @anttin2020 Wonderful answer. Million thanks! I'm gonna give it a go.

Thank you :)

It may look confusing because of that path mess but I tried to use some arbitrary folder for the edited source and that did not work so that a bit messy method worked for me. If you have problems, let me know and I'll try to help.

Also I suggest you export the SpaceMouse settings from Editor settings in 5.1 and then import them later in 5.2, that way you will get exactly same settings directly. Of course if you haven't changed anything then that is not needed.

Good luck!

And once again big thank you to @microdee , this plugin is one of the most important ones in UE, it makes everything related to navigating so easy and also gives Take Recorder a whole new meaning!

**Comment by MarkEHawkins** [comment]
Hi, thanks for help. Can anyone tell me what I'm doing wrong?
**'RunUAT.bat'' is not recognized as an internal or external command, operable program or batch file.**
I have visual studio 2019 and C++ library installed. Am I missing a step? 
Added #include "SceneView.h" to file SmViewportOverlay.cpp,

cd to my batch folder
RunUAT.bat BuildPlugin -plugin="D:\_UnrealEngine _Engine\UE_5.2\Engine\Plugins\spacemouse\SpaceMouse.uplugin -package="D:\newplugin" -TargetPlatforms=Win64

Thank you

**Comment by anttin2020** [comment]
@MarkEHawkins 
See my instructions step3, you need to be in S:\UE\UE_5.2\Engine\Build\BatchFiles :)
That batch file is in that folder and as it is not on system path, it cannot be found anywhere else.

Good luck!

**Comment by MarkEHawkins** [comment]
Hi, thanks for your help.  Do I need 5.1 to be installed? 
Step 3. I have cd into d dive then to my D:\_UnrealEngine\UE_5.2\Engine\Build\BatchFiles _Engine\UE_5.2\Engine\Build\BatchFiles>RunUAT.bat BuildPlugin -plugin="D:\_UnrealEngine _Engine\UE_5.2\Engine\Plugins\spacemouse\SpaceMouse.uplugin -package="D:\newplugin"
Running AutomationTool...
Using bundled DotNet SDK version: 6.0.302
Starting AutomationTool...
Parsing command line: BuildPlugin -plugin="D:\_UnrealEngine _Engine\UE_5.2\Engine\Plugins\spacemouse\SpaceMouse.uplugin -package=D:\newplugin"
Initializing script modules...
Total script module initialization time: 1.87 s.
Executing commands...
ERROR: Plugin 'D:\_UnrealEngine _Engine\UE_5.2\Engine\Plugins\spacemouse\SpaceMouse.uplugin -package=D:\newplugin' not found
     
AutomationTool executed for 0h 0m 2s
AutomationTool exiting with ExitCode=1 (Error_Unknown)
BUILD FAILED

**Comment by anttin2020** [comment]
@MarkEHawkins 
You do need working version of the plugin for 5.1 but the one from here should be just that.
But the paths need to be absolute so you need to use D:\_UnrealEngine _Engine\UE_5.2 (note the missing \ on your version)

**Comment by MarkEHawkins** [comment]
Hi, Thanks, Sorry, I just removed the TargetPlatforms for the last try. I will add it again. 
`RunUAT.bat BuildPlugin -plugin="D:\_UnrealEngine _Engine\UE_5.2\Engine\Plugins\spacemouse\SpaceMouse.uplugin -package="D:\newplugin" -TargetPlatforms=Win64`

**Comment by anttin2020** [comment]
You are missing trailing " for the plugin path :)

**Comment by MarkEHawkins** [comment]
Thank you so much for your help. 
I have been going round in circles over the last week trying to build the 5.2 plugin. I used this page for reference without the ". for the plugin build [https://dev.epicgames.com/community/learning/tutorials/0RG4/compiling-a-plugin-for-a-new-engine-version](url).
I nearly gave up then you came to the rescue with your fix but I still kept the old file path to the SpaceMouse.uplugin.
I was code blind this morning!!
Thanks for pouting that error.

**Comment by anttin2020** [comment]
No problem, great that you got it to work, now you can enjoy UE 5.2 with 3D mouse as well which is a must, I couldn't think using it without anymore :)

**Comment by MarkEHawkins** [comment]
Thanks very much. 
I had to install NET 4.6, did not compile. After install now compiled. 

Sorry to be a pain. Compiling the plugin, I had a lot of yellow warnings!
"Calling this function through the result pointer may cause your program to fail" Also  "warning C4191: 'type cast': unsafe conversion from 'FARPROC' to 'HidD_FreePreparsedData_'"
Installed it and it works ok so far. Do you get yellow warnings and should I just ignore it?
Are you able to share your 3dmouse user setting? It would really help to see how other users have setup the speed etc. So far I just delete the minus (-)in each Axis and then add the minus(-) where there was not a minus.

**Comment by anttin2020** [comment]
@MarkEHawkins 
Sorry I didn't notice the reply until now.
No pain at all, I'm happy to help if I just can :)

I got bunch of warnings as well but as the build was successful, I ignored them and everything seems to work so far so I think it is safe to ignore them :)

Here is my spacemouse settings (had to zip as git does not accept ini attachments) but I only use it for camera/viewport navigation so I'm afraid there is nothing special for you. I also have not used Sketchup so I don't know how it works and those minus things you mentioned are unknown to me.

[editor prefs_SpaceMouse Backup 2023-01-03 092310.zip](https://github.com/microdee/UE4-SpaceMouse/files/11642932/editor.prefs_SpaceMouse.Backup.2023-01-03.092310.zip)

**Comment by MarkEHawkins** [comment]
Thanks, that's good to know warnings are not causing any issues for you too.. 
Thanks for sending your settings. The rotation settings work better than my previous setup. I have to add the minus thing in UE5 or everything is the opposite (up is down down is up etc). See attached, pink is your setting. I'm not sure what is different with my setup (I'm not balancing on my head ;D) 
![image](https://github.com/microdee/UE4-SpaceMouse/assets/1959647/cde7491a-aaf0-4200-a6f6-f53a61781fa9)

**Comment by anttin2020** [comment]
I don't remember changing anything in those, with 3Dconnexion driver I had to swap the directions but with this I don't remember needing to do that. That does not prove anything, Jan was long time ago and I remember tweaking the settings for a while so... :)

But if those settings work for you too, then nice! No matter what is the differerence unless for curiosity ;)

**Comment by MarkEHawkins** [comment]
I'm very happy now with the plugin is working and settings close to what I'm used to. Nothing seems broken so I will just leave it alone. Its exciting to get into 5.2. Thanks AnttiN and David Mórász.

**Comment by microdee** [comment]
Sorry about the late response, but finally I got to make the necessary updates and there's a release now for 5.2 https://github.com/microdee/UE4-SpaceMouse/releases/latest

**Comment by anttin2020** [comment]
@microdee 
Thank you! Did you change something else than added the include i.e. should we take your update or is the one built from modified sources ok? 
Are the yellow warnings (see above) ok or did we do something wrong? Everything seems to work but just to make sure :)

**Comment by microdee** [comment]
I haven't added any other functionality. Besides the includes, I fixed some build warnings and complied with deprecation notices + updated my build tool.

**Comment by anttin2020** [comment]
Ok, then I will wait until you update the Marketplace version and then buy this to support your excellent work. I wanted to buy this earlier but there was no proper version for current UE so I decided to go with git but now that I have seen how well it works also in UE5.2 with mod sources, I want to buy it as well :)
So please update the Marketplace version soon :)

**Comment by MarkEHawkins** [comment]
Hi Just deleted my working 5.2 version 5.2 for v1.26. Added it to marketplace folder but now getting the error!
![image](https://github.com/microdee/UE4-SpaceMouse/assets/1959647/77880b4c-a986-471e-a552-25135f30eb69)

**Comment by microdee** [comment]
oh that's not good, I'm looking into it

**Comment by MarkEHawkins** [comment]
Thanks. No rush on my part. Remade my previous build and working again.

**Comment by anttin2020** [comment]
I noticed a weird problem: when I am in editor, the +/- speed buttons adjust the viewport camera speed as expected. Then I go to play mode and of course the speed buttons do not do anything as is (without events in BP) BUT when I return from play mode, the viewport camera speed has changed :o
Is there any way to prevent the editor camera speed from changing 'secretly' while in play mode?

I am still using my own compiled version as there is no confirmation that the Marketplace version is fixed but as I have not changed anything I believe the current version behaves the same :)

Any ideas?

**Comment by microdee** [comment]
This was supposedly fixed some time ago with 1.2.9 release so I'm closing this

**Comment by anttin2020** [comment]
I wanted to try to use the runuat.bat to recompile this for UE 5.4 but it seems the whole runuat.bat is broken :o
Here's what I got:
```
S:\UE\UE_5.4\Engine\Build\BatchFiles>RunUAT.bat BuildPlugin-plugin="s:\\ue\\ue_5.3\\engine\\plugins\\spacemouse\\spacemouse.uplugin"-package="s:\\newplugin" -TargetPlatforms=Win64
Running AutomationTool...
Using bundled DotNet SDK version: 6.0.302
Starting AutomationTool...
Parsing command line: BuildPlugin-plugin=s:\\ue\\ue_5.3\\engine\\plugins\\spacemouse\\spacemouse.uplugin-package=s:\\newplugin -TargetPlatforms=Win64
ERROR: Unknown parameter -TargetPlatforms=Win64 in the command line that does not belong to any command.
BUILD FAILED
```
Has someone managed to rebuild this or any plugin for UE 5.4? Is it broken and if not, what is the correct command line now?
I also tried to use -platform instead of -TargetPlatforms but same error. I haven't found any documentation for this, only old forum posts that of course give the old method.

Please help :)

@microdee, please make this excellent plugin to work in UE 5.4, the UE 3DConnexion plugin is only partially usable as it cannot work in runtime.
